crates/ruff_python_semantic/src/model.rs

```diff
@@ -36,8 +36,11 @@ pub struct SemanticModel<'a> {
     /// The identifier of the current statement.
     stmt_id: Option<NodeId>,
 
-    /// Stack of current expressions.
-    exprs: Vec<&'a Expr>,
+    /// Stack of all visited expressions.
+    exprs: Nodes<'a, Expr>,
+
+    /// The identifier of the current expression.
+    expr_id: Option<NodeId>,
```

Comment:
Nit: We now store both a stack of statements and expressions. This gives us almost a full fidelity parent chain, except that it isn't possible to get the `parent` statement of an expression (except for the current expression). 

What do you think of introducing a `Vec<AnyNodeRef>` that stores all parent nodes (Ultimately means `O(n)` pointers to be written, where `n` = len(AST)) instead? You can retrieve the current statement by traversing upwards until you find the first `is_statement()` node. 

The one issue of that is that there's currently no way to go from `AnyNodeRef` to `&Stmt`. We would need to introduce a `AnyStatementRef` enum for that.

Reply:
I will consider this separately, it makes sense to me, although it may come with the performance regression that I documented here and fixed in https://github.com/astral-sh/ruff/pull/6351.

Reply:
I was looking into this and I want to understand the last paragraph here. Is the suggestion that methods like `SemanticModel#current_statement` would now return `AnyStatementRef`? And so we'd change all consumers of that method to work with `AnyStatementRef` rather than `&Stmt`?

Reply:
If so, we also need `AnyExpressionRef`, and I'd ultimately like to store the full-fidelity chain here including `Parameters`, `Arguments`, etc...

Reply:
E.g., these kinds of changes: https://github.com/astral-sh/ruff/compare/charlie/any-node-ref?expand=1#diff-c2c8422959405016e1fb54ca23edee9716bd06ad64e610ceea9612a8291122fcR1815. It's feasible but a lot of call sites to work through, so I want to make sure I understand what's being suggested before I think on it further.

Reply:
I might propose that we do this incrementally by adding an internal type to the semantic crate that's like `AnyNodeRef`, but less granular:

```rust
#[derive(Debug, Clone)]
pub enum AnyNodeRef<'a> {
    Stmt(&'a Stmt),
    Expr(&'a Expr),
}

impl<'a> AnyNodeRef<'a> {
    pub fn as_statement(&self) -> Option<&'a Stmt> {
        match self {
            AnyNodeRef::Stmt(stmt) => Some(stmt),
            AnyNodeRef::Expr(_) => None,
        }
    }
}
```

That way, we wouldn't need to change the `SemanticModel` interface at all for now.